### PR TITLE
Add git-version task

### PR DESCRIPTION
**Pull request by GijsvanDulmen**
# Changes

Add a task to generate versions based on git history.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [X] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [X] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [X] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [X] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [X] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [X] mandatory `spec.description` follows the convention

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

File: task/git-version/0.1/git-version.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: git-version
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.21.0"
```
Comment (vinamra28) [inline task/git-version/0.1/git-version.yaml:9]: any reason for setting `pipelines.minVersion` to `0.21.0` ?

File: task/git-version/0.1/git-version.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    - description: branch to checkout to create a version for
+      name: branch
+      type: string
+  results:
+    - description: The calculated git version you could use for git tagging
+      name: gitVersion
+    - description: A normalized version for use in container images for example
+      name: packageVersion
+  steps:
+    - image: "mcr.microsoft.com/dotnet/sdk:3.1-focal"
```
Comment (vinamra28) [inline task/git-version/0.1/git-version.yaml:26]: can we tag the image with digest as well?

**Review by vinamra28 — COMMENTED** [review]
can you please add a test for this task as well?

File: task/git-version/0.1/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+## Generate a version from git history
+
+The `git-version` task let you generate a version from git history using `https://gitversion.net/`
+
+### Install the Task
+
+```
+kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/github-version/0.1/git-version.yaml
```
Comment (vinamra28) [inline task/git-version/0.1/README.md:8]: nit:
```suggestion
kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/git-version/0.1/git-version.yaml
```

**Review by vinamra28 — APPROVED** [review]
/meow

File: task/git-version/0.1/tests/pre-apply-task-hook.sh [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+# Add git-clone
+add_task git-cli latest
```
Comment (piyush-garg) [inline task/git-version/0.1/tests/pre-apply-task-hook.sh:4]: comment above script mentions git clone, and git-cli task is applied here

Comment (piyush-garg) [inline task/git-version/0.1/tests/pre-apply-task-hook.sh:4]: comment above script mentions git clone, and git-cli task is applied here

File: task/git-version/0.1/tests/run.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+kind: Pipeline
+metadata:
+  name: git-version-test
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - name: git-fetch-all
+      taskRef:
+        name: git-cli
```
Comment (piyush-garg) [inline task/git-version/0.1/tests/run.yaml:12]: cant we use git-clone task here

File: task/git-version/0.1/git-version.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  description: >-
+    This task can be used to create a version from git history
+  params:
+    - description: branch to checkout to create a version for
+      name: branch
+      type: string
+  results:
+    - description: The calculated git version you could use for git tagging
+      name: gitVersion
+    - description: A normalized version for use in container images for example
```
Comment (piyush-garg) [inline task/git-version/0.1/git-version.yaml:23]: can you remove `for example`

File: task/git-version/0.1/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+* **source**: A workspace with a checked out git repository.
+
+### Parameters
+
+* **branch:**: The branch to checkout and generate a version for.
+
+## Results
+
+* **gitVersion**: The calculated git version you could use for git tagging
```
Comment (piyush-garg) [inline task/git-version/0.1/README.md:21]: can you provide eg for parameters and results

File: task/git-version/0.1/tests/run.yaml (hunk 5 above)
Comment (GijsvanDulmen) [reply task/git-version/0.1/tests/run.yaml:12]: No, git-clone doesn't "fetch all" branches which is needed for the git version tool.

**Comment by piyush-garg** [comment]
@GijsvanDulmen Can you please provide the information about platforms if possible using the annotation

**Comment by GijsvanDulmen** [comment]
> @GijsvanDulmen Can you please provide the information about platforms if possible using the annotation

Done!

**Comment by piyush-garg** [comment]
/lgtm

Thanks for PR

**Pull request merged**